notation.go

```diff
@@ -63,31 +45,202 @@ type SignOptions struct {
+}
+
+// Verify performs signature verification on each of the notation supported
+// verification types (like integrity, authenticity, etc.) and return the
+// verification outcomes.
+// For more details on signature verification, see
+// https://github.com/notaryproject/notaryproject/blob/main/specs/trust-store-trust-policy.md#signature-verification
+func Verify(ctx context.Context, verifier Verifier, repo registry.Repository, opts VerifyOptions) (ocispec.Descriptor, []*VerificationOutcome, error) {
+	// passing nil signature to check 'skip'
+	_, outcome, err := verifier.Verify(ctx, nil, opts)
+	if err != nil {
+		if outcome == nil {
+			return ocispec.Descriptor{}, nil, err
+		}
+	} else if reflect.DeepEqual(outcome.VerificationLevel, trustpolicy.LevelSkip) {
+		return ocispec.Descriptor{}, []*VerificationOutcome{outcome}, nil
+	}
+	// get signature manifests
+	artifactRef := opts.ArtifactReference
+	artifactDescriptor, err := repo.Resolve(ctx, artifactRef)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, ErrorSignatureRetrievalFailed{Msg: err.Error()}
+	}
+
+	var verificationOutcomes []*VerificationOutcome
+	var targetArtifactDesc ocispec.Descriptor
+	if opts.MaxVerificationLimit <= 0 {
+		// Set MaxVerificationLimit to 50 as default
+		opts.MaxVerificationLimit = maxVerificationLimitDefault
+	}
+	count := 0
+	err = repo.ListSignatures(ctx, artifactDescriptor, func(signatureManifests []ocispec.Descriptor) error {
+		// process signatures
+		for _, sigManifestDesc := range signatureManifests {
+			if count >= opts.MaxVerificationLimit {
+				break
+			}
+			count++
+			// get signature envelope
+			sigBlob, _, err := repo.FetchSignatureBlob(ctx, sigManifestDesc)
+			if err != nil {
+				return ErrorSignatureRetrievalFailed{Msg: fmt.Sprintf("unable to retrieve digital signature with digest %q associated with %q from the registry, error : %v", sigManifestDesc.Digest, artifactRef, err.Error())}
+			}
+			payloadArtifactDescriptor, outcome, err := verifier.Verify(ctx, sigBlob, opts)
+			if err != nil {
+				if outcome == nil {
+					// TODO: log fatal error
+					return err
+				}
+				verificationOutcomes = append(verificationOutcomes, outcome)
+				continue
+			}
+			if !content.Equal(payloadArtifactDescriptor, artifactDescriptor) {
+				outcome.Error = errors.New("content descriptor mismatch")
+				verificationOutcomes = append(verificationOutcomes, outcome)
+				continue
+			}
+
+			// At this point, we've found a signature verified successfully
+			verificationOutcomes = append(verificationOutcomes, outcome)
+			// Descriptor of the signature blob that gets verified successfully
+			targetArtifactDesc = payloadArtifactDescriptor
+
+			return errDoneVerification
+		}
+
+		if count >= opts.MaxVerificationLimit {
+			return ErrorVerificationFailed{Msg: fmt.Sprintf("number of signatures associated with an artifact should be less than: %d", opts.MaxVerificationLimit)}
```

Comment:
When it reaches the max verification limit should we return the previous outcomes along with the above error?